### PR TITLE
Eng 1652 pop out http errors to new npm repo

### PR DESCRIPTION
**Pull request by shellylane**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,15 @@
-# http-errors
\ No newline at end of file
+# Custom HTTP Errors
+
+To install for development:
+`yarn add ../http-errors`
```
Comment (wtkersting) [inline README.md]: Remove this

File: index.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  409: 'conflict',
+  418: 'i am a teapot',
+  420: 'enhance your calm',
+  422: 'unprocessable entity',
+  500: 'internal server error',
+}
+
+/**
+ * Generic HTTP errors have JSON message and status
+ */
```
Comment (wtkersting) [inline index.js]: Can you add `@typedef`s?

**Review by wtkersting — CHANGES_REQUESTED** [review]

File: index.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+class HttpError extends Error {
+  /**
+   *
```
Comment (wtkersting) [inline index.js]: Let's keep this file clean - we can remove these empty lines for comment blocks

File: index.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+/**
+ * Generic HTTP errors have JSON message and status
+ */
+class HttpError extends Error {
+  /**
+   * 
+   * @param {number} statusCode 
+   * @param {string} message 
+   * @param {string} externalMessage
```
Comment (wtkersting) [inline index.js]: externalMessage can also be a `boolean`

**Review by wtkersting — CHANGES_REQUESTED** [review]

File: index.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
+// translate status code to response message
+const response = {
```
Comment (daverted) [inline index.js]:
```suggestion
/** @type {{[key: number]: string}} */
const response = {
```

File: index.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  409: 'conflict',
+  418: 'i am a teapot',
+  420: 'enhance your calm',
+  422: 'unprocessable entity',
+  500: 'internal server error',
+}
+
+/**
+ * Generic HTTP errors have JSON message and status
+ */
```
Comment (daverted) [inline index.js]:
```suggestion
 * @extends {Error}
 */
```

File: index.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    this.externalMessage = externalMessage === true ? message : externalMessage
+    let json = { error: response[statusCode], statusCode: statusCode }
+    this.json = externalMessage ? { ...json, message: this.externalMessage } : json
+    this.response = JSON.stringify(this.json)
+  }
+}
+
+/**
+ * Custom error denoting a bad request was made (HTTP 400)
+ */
```
Comment (daverted) [inline index.js]:
```suggestion
 * @extends {HttpError}
 */
```

**Review by daverted — CHANGES_REQUESTED** [review]
Take a look at https://jsdoc.app/ and add proper typedefs. I've added a few in the comments to get you started.

File: typedoc.mjs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+  const project = app.convert()
+
+  if (project) {
+    // project may not have converted correctly
+    await app.generateDocs(project, 'docs')
+  }
+}
+
+main().catch(console.error)
```
Comment (jonnysowards) [inline typedoc.mjs]: Add a line at the end of the file.

```suggestion
main().catch(console.error)

```

File: typedoc.json [hunk 9]
```diff
@@ -0,0 +1,7 @@
+{
+    "pretty": true,
+    "includeVersion": true,
+    "disableSources": true,
+    "cleanOutputDir": true,
+    "theme": "markdown"
+  }
```
Comment (jonnysowards) [inline typedoc.json]: Add new line at end of file.

```suggestion
  }
  
```

File: tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+      "strict": true,
+      "declarationDir": "./types",
+      "declaration": true,
+      "target": "es6",
+      "alwaysStrict": true,
+      "composite": true
+    },
+    "include": ["src/*.js"],
+    "exclude": ["node_modules"]
+  }
```
Comment (jonnysowards) [inline tsconfig.json]: Add new line at end of file.
```suggestion
  }
  
```

**Review by jonnysowards — CHANGES_REQUESTED** [review]

**Review by jonnysowards — APPROVED** [review]

File: .github/workflows/publish.yaml [hunk 11]
```diff
@@ -59,4 +59,49 @@ jobs:
...
+          rsync -r ./library/docs/ ./docs
+
+      - name: Push the updated docs
+        working-directory: ./docs
+        run: |
+          git config --local user.email 'action@github.com'
+          git config --local user.name 'GitHub Action'
+          git add .
+          git commit -m 'Update docs for library version ${{ needs.bump.outputs.newTag }}'
+          git push
```
Comment (wtkersting) [inline .github/workflows/publish.yaml:107]: Add a new line

**Review by wtkersting — APPROVED** [review]
Approved, though make sure that NPM is ready to publish before merging

File: .github/workflows/publish.yaml (hunk 11 above)
Comment (wtkersting) [reply .github/workflows/publish.yaml:107]: There's still not a new line here

File: README.md [hunk 12]
```diff
@@ -1 +1,12 @@
-# http-errors
\ No newline at end of file
+# Custom HTTP Errors
+
+To install for production:
```
Comment (wtkersting) [inline README.md]:
```suggestion
To install
```

File: src/index.js [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+/**
+ * Generic HTTP errors have JSON message and status
+ * @class
+ * @extends {Error}
+ */
+class HttpError extends Error {
+  /**
+   * @param {number} statusCode 
+   * @param {string} message 
+   * @param {string | boolean} externalMessage
```
Comment (wtkersting) [inline src/index.js]: All of these `externalMessage` properties are optional - they need to be in `[]` -
```suggestion
   * @param {string | boolean} [externalMessage]
```

image

**Review by wtkersting — CHANGES_REQUESTED** [review]
Make the `externalMessage` properties `[optional]`

**Review by wtkersting — APPROVED** [review]

**Review by daverted — APPROVED** [review]

**Pull request merged**